lib/Platform/Unicode/CharacterProperties.cpp

```diff
@@ -210,4 +211,122 @@ uint32_t canonicalize(uint32_t cp, bool unicode) {
   }
 }
 
+#ifdef HERMES_ENABLE_UNICODE_REGEXP_PROPERTY_ESCAPES
+
+/// Find a matching entry (such as \p NameMapEntry or \p RangeMapEntry) by
+/// matching a string \p name against the entry's \p name field.
+template <class T>
+const T *findNameMapEntry(
+    const llvh::ArrayRef<T> &arrayRef,
+    const std::string_view name) {
+  auto it = std::lower_bound(
+      std::begin(arrayRef),
+      std::end(arrayRef),
+      name,
+      [](const T &a, std::string_view b) {
+        return UNICODE_DATA_STRING_POOL.compare(a.name.offset, a.name.size, b) <
+            0;
+      });
+  if (it == std::end(arrayRef) ||
+      UNICODE_DATA_STRING_POOL.compare(it->name.offset, it->name.size, name) !=
+          0) {
+    return nullptr;
+  }
+  return it;
+}
+
+bool addUnicodePropertyRanges(
+    CodePointSet *receiver,
+    std::string_view propertyName,
+    std::string_view propertyValue,
+    bool inverted = false) {
+  auto key = propertyName;
+  llvh::ArrayRef<RangeMapEntry> rangeMap;
+  llvh::ArrayRef<NameMapEntry> nameMap;
+
+  if (propertyValue.empty()) {
+    rangeMap = unicodePropertyRangeMap_BinaryProperty;
+    nameMap = canonicalPropertyNameMap_BinaryProperty;
+    // There was no property value, this is either a binary property or a value
+    // from General_Category, as per `LoneUnicodePropertyNameOrValue`.
+    if (findNameMapEntry(nameMap, key) == nullptr) {
+      rangeMap = unicodePropertyRangeMap_GeneralCategory;
+      nameMap = canonicalPropertyNameMap_GeneralCategory;
+    }
+  } else {
+    // There was a property value, assume the name is a category.
+    key = propertyValue;
+    if (propertyName == "General_Category" || propertyName == "gc") {
+      rangeMap = unicodePropertyRangeMap_GeneralCategory;
+      nameMap = canonicalPropertyNameMap_GeneralCategory;
+    } else if (propertyName == "Script" || propertyName == "sc") {
+      rangeMap = unicodePropertyRangeMap_Script;
+      nameMap = canonicalPropertyNameMap_Script;
+    } else if (propertyName == "Script_Extensions" || propertyName == "scx") {
+      // Script_Extensions is a superset of Script.
+      if (!addUnicodePropertyRanges(receiver, "Script", propertyValue)) {
```

Comment:
Does this mean that the `Script_Extensions` ranges encoded only contain the intersection of the ranges for `Script_Extensions` and `Script`? Could we lay out the data such that this isn't the case?

(Calling this out because I realise this might complicate the proposal I made about using an `ArrayRef` in `ClassAtom`)